common/nexus/failure.go

```diff
@@ -47,65 +51,124 @@ func ProtoFailureToNexusFailure(failure *nexuspb.Failure) *nexus.Failure {
 
 // NexusFailureToProtoFailure converts a Nexus SDK Failure to a proto Nexus Failure.
 // Always returns a non-nil value.
-func NexusFailureToProtoFailure(failure *nexus.Failure) *nexuspb.Failure {
-	if failure == nil {
-		return &nexuspb.Failure{Message: "unknown error"}
-	}
+func NexusFailureToProtoFailure(failure nexus.Failure) *nexuspb.Failure {
 	return &nexuspb.Failure{
 		Message:  failure.Message,
 		Metadata: failure.Metadata,
 		Details:  failure.Details,
 	}
 }
 
-// APIFailureToNexusFailure converts an API proto Failure to a Nexus SDK Failure taking only the failure message to
-// avoid leaking too many details to 3rd party callers.
+// APIFailureToNexusFailure converts an API proto Failure to a Nexus SDK Failure setting the metadata "type" field to
+// the proto fullname of the temporal API Failure message.
 // Always returns a non-nil value.
-func APIFailureToNexusFailure(failure *failurepb.Failure) *nexus.Failure {
-	return &nexus.Failure{
-		Message: failure.GetMessage(),
+func APIFailureToNexusFailure(failure *failurepb.Failure) (nexus.Failure, error) {
+	// Unset message so it's not serialized in the details.
+	failureCopy := common.CloneProto(failure)
+	failureCopy.Message = ""
+	data, err := protojson.Marshal(failureCopy)
+	if err != nil {
+		return nexus.Failure{}, err
 	}
```

Comment:
Why not just save the message and restore afterwards instead of cloning the entire proto?
```suggestion
	failureMessage := failure.GetMessage()
	failure.Message = ""
	data, err := protojson.Marshal(failure)
	if err != nil {
		return nexus.Failure{}, err
	}
	failure.Message = failureMessage
```

Reply:
I was being a bit conservative here but I guess we can be make an assumption that these failure protos won't be accessed concurrently and mutate the value passed in.